velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -1291,13 +1292,22 @@ std::pair<std::unique_ptr<TaskCursor>, std::vector<RowVectorPtr>> readCursor(
     addSplits(task);
   }
 
-  EXPECT_TRUE(waitForTaskCompletion(task, maxWaitMicros)) << task->taskId();
+  if (!waitForTaskCompletion(task, maxWaitMicros)) {
+    // NOTE: there is async memory arbitration might fail the task after all the
+    // results have been consumed and before the task finishes. So we might run
+    // into the failed task state in some rare case such as exposed by
+    // concurrent memory arbitration test.
+    if (task->state() != Task::State::kFinished) {
+      waitForTaskDriversToFinish(task);
```

Comment:
Do we need to make sure `task->error()` is true here, and add a timeout parameter?

Reply:
task->error() is always set if task has failed. I updated the code to throw a different error if the task hasn't failed. Thanks for pointing out.